### PR TITLE
Feature thumbnail image decoding

### PR DESCRIPTION
**Pull request by dreampiggy**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass // TODO
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: #2810

### New APIs

New decoding options:

+ SDImageCoderDecodePreserveAspectRatio

Control whether to follow aspect ratio. Defaults to YES to match most user's usage.

+ SDImageCoderDecodeThumbnailPixelSize:

Control the max thumbnail pixel size. CGSize value.

This can also be used for vector format coder, to replace such as [SDWebImageContextPDFImageSize](https://github.com/SDWebImage/SDWebImagePDFCoder/blob/0.5.1/SDWebImagePDFCoder/Classes/SDWebImagePDFCoderDefine.h)

Both of the options, support the progressive decoding as well as animated decoding, which means each frame will use the thumbnail image and save memory.

### Notable changes

This `SDWebImageScaleDownLargeImages` option, previouslly use the complicated draw code to draw smaller bitmap which limit to the 60MB bytes. Now, it's translated to use the thumbnail decoding as well. This may be a visible changes, but most cases, this should just works.

### CODE REVIEW

File: SDWebImage/Core/SDImageCoderHelper.h [hunk 1]
```diff
@@ -90,6 +90,12 @@
 + (UIImage * _Nullable)decodedAndScaledDownImageWithImage:(UIImage * _Nullable)image limitBytes:(NSUInteger)bytes;
 
 #if SD_UIKIT || SD_WATCH
+/**
+ Control the default limit bytes to scale down larget images.
+ This value must be larger than or equal to 1MB. Defaults to 60MB.
+ */
+@property (class, readwrite) NSUInteger defaultScaleDownLimitBytes;
+
```
Comment (dreampiggy) [inline SDWebImage/Core/SDImageCoderHelper.h]: One strange question:
Why all the force decode method (`decodedImageWithImage` and `decodedAndScaledDownImageWithImage:limitBytes:`) only applied on iOS/tvOS/watchOS ?
Those methods use Core Graphics, and macOS have the same API as well. macOS still use Core Animation for rendering, so I don't think they should be ignored.

Comment (dreampiggy) [reply SDWebImage/Core/SDImageCoderHelper.h]: What about we change those method to let it works on macOS as well.

The default limit bytes on macOS can be doubled, like 120MB (MacBook's RAM is 4GB+)

File: SDWebImage/Core/SDImageIOAnimatedCoder.m [hunk 2]
```diff
@@ -265,20 +339,8 @@ - (UIImage *)incrementalDecodedImageWithOptions:(SDImageCoderOptions *)options {
...
-                scale = MAX([scaleFactor doubleValue], 1);
-            }
-#if SD_UIKIT || SD_WATCH
-            image = [[UIImage alloc] initWithCGImage:partialImageRef scale:scale orientation:UIImageOrientationUp];
-#else
-            image = [[UIImage alloc] initWithCGImage:partialImageRef scale:scale orientation:kCGImagePropertyOrientationUp];
-#endif
-            CGImageRelease(partialImageRef);
+        image = [SDImageIOAnimatedCoder createFrameAtIndex:0 source:_imageSource scale:_scale preserveAspectRatio:_preserveAspectRatio thumbnailSize:_thumbnailSize];
+        if (image) {
```
Comment (kinarobin) [inline SDWebImage/Core/SDImageIOAnimatedCoder.m]: Maybe you had forgot `SDImageCoderDecodeScaleFactor `?

File: SDWebImage/Core/SDImageCoderHelper.m [hunk 3]
```diff
@@ -311,13 +300,11 @@ + (UIImage *)decodedAndScaledDownImageWithImage:(UIImage *)image limitBytes:(NSU
...
     CGFloat tileTotalPixels;
-    if (bytes > 0) {
-        destTotalPixels = bytes / kBytesPerPixel;
-        tileTotalPixels = destTotalPixels / 3;
-    } else {
-        destTotalPixels = kDestTotalPixels;
-        tileTotalPixels = kTileTotalPixels;
+    if (bytes == 0) {
+        bytes = kDestImageLimitBytes;
     }
```
Comment (kinarobin) [inline SDWebImage/Core/SDImageCoderHelper.m]: How about `bytes <= 0` ?

Comment (dreampiggy) [reply SDWebImage/Core/SDImageCoderHelper.m]: bytes is `NSUInteger`

File: SDWebImage/Core/SDImageIOCoder.m [hunk 4]
```diff
@@ -140,21 +185,8 @@ - (UIImage *)incrementalDecodedImageWithOptions:(SDImageCoderOptions *)options {
...
-            }
-#if SD_UIKIT || SD_WATCH
-            UIImageOrientation imageOrientation = [SDImageCoderHelper imageOrientationFromEXIFOrientation:_orientation];
-            image = [[UIImage alloc] initWithCGImage:partialImageRef scale:scale orientation:imageOrientation];
-#else
-            image = [[UIImage alloc] initWithCGImage:partialImageRef scale:scale orientation:_orientation];
-#endif
-            CGImageRelease(partialImageRef);
+        image = [SDImageIOAnimatedCoder createFrameAtIndex:0 source:_imageSource scale:_scale preserveAspectRatio:_preserveAspectRatio thumbnailSize:_thumbnailSize];
+        if (image) {
```
Comment (kinarobin) [inline SDWebImage/Core/SDImageIOCoder.m]: Same to  `SDImageCoderDecodeScaleFactor `

File: SDWebImage/Core/SDImageIOAnimatedCoder.m (hunk 2 above)
Comment (dreampiggy) [reply SDWebImage/Core/SDImageIOAnimatedCoder.m]: @kinarobin This is the progressive decoding. The `_scale` is initialized in `initIncrementalWithOptions:` method. See above 281line.

Comment (dreampiggy) [reply SDWebImage/Core/SDImageIOAnimatedCoder.m]: Oh....I forgot this edge case. If you pass `SDImageCoderDecodeScaleFactor ` in both `initIncrementalWithOptions:`, and also this `incrementalDecodedImageWithOptions:`. We prefer to use this method arg firstlly, then backup with `initIncrementalWithOptions:`... Fix it

File: SDWebImage/Core/SDImageCoderHelper.h (hunk 1 above)
Comment (kinarobin) [reply SDWebImage/Core/SDImageCoderHelper.h]: That's a good idea!

**Comment by KittenYang** [comment]
Please have a look to this discussion: https://github.com/SDWebImage/SDWebImage/discussions/3192

**Pull request merged**